platforms/software/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultLocalComponentGraphResolveState.java

```diff
@@ -273,4 +369,108 @@ public AttributesSchemaInternal getAttributesSchema() {
             return metadata.getAttributesSchema();
         }
     }
+
+    private static class DefaultLocalComponentGraphSelectionCandidates implements LocalComponentGraphSelectionCandidates {
+        private final List<? extends VariantGraphResolveState> variantsWithAttributes;
+        private final Map<String, ? extends VariantGraphResolveState> variantsByConfigurationName;
+        private final AbstractComponentGraphResolveState<?> component;
+
+        public DefaultLocalComponentGraphSelectionCandidates(
+            List<? extends VariantGraphResolveState> variantsWithAttributes,
+            Map<String, ? extends VariantGraphResolveState> variantsByConfigurationName,
+            AbstractComponentGraphResolveState<?> component
+        ) {
+            this.variantsWithAttributes = variantsWithAttributes;
+            this.variantsByConfigurationName = variantsByConfigurationName;
+            this.component = component;
+        }
+
+        @Override
+        public boolean supportsAttributeMatching() {
+            return !variantsWithAttributes.isEmpty();
+        }
+
+        @Override
+        public List<? extends VariantGraphResolveState> getVariantsForAttributeMatching() {
+            if (variantsWithAttributes.isEmpty()) {
+                throw new IllegalStateException("No variants available for attribute matching.");
+            }
+            return variantsWithAttributes;
+        }
+
+        @Nullable
+        @Override
+        public VariantGraphResolveState getVariantByConfigurationName(String name) {
+            VariantGraphResolveState variant = variantsByConfigurationName.get(name);
+            if (variant != null) {
+                return variant;
+            }
+
+            // There is no consumable variant with the given name.
+            // Perhaps there is a configuration with the same name, but it is not consumable.
+            // In that case, throw an error.
+            ConfigurationGraphResolveState conf = component.getConfiguration(name);
+            if (conf == null) {
+                return null;
+            }
+
+            // If the configuration exists, it must not be consumable, since variantsByConfigurationName contains
+            // all consumable configurations.
+            assert !conf.getMetadata().isCanBeConsumed() : "Expected configuration to be non-consumable";
+
+            ConfigurationNotConsumableFailure failure = new ConfigurationNotConsumableFailure(name, component.getId().getDisplayName());
+            String message = String.format("Selected configuration '" + failure.getRequestedName() + "' on '" + failure.getRequestedComponentDisplayName() + "' but it can't be used as a project dependency because it isn't intended for consumption by other components.");
+            throw new ConfigurationSelectionException(message, failure, Collections.emptyList());
+        }
```

Comment:
This is what I'd like to discuss the most

I'm not sure I really see the value of using a failure handler here, since I can't imagine any situation where there would need to be any custom handling done here. I feel we add more complexity to the handling of errors like this by needing to wire a resolution handler to the failure location, implementing multiple methods on the handler, implementing a failure describer, and registering that with the handler. 

I can see the value for errors related to attributes, where there are semantics behind the attributes that may warrant custom messages, but I thought from prior discussions we agreed we woudln't want to expose similar functionality for non-attribute related failures.

Reply:
> I thought from prior discussions we agreed we woudln't want to expose similar functionality for non-attribute related failures

I don't remember concluding this.  The intent was to handle all resolution failures during selection.  We already support selection failures due to bad configuration names when using explicit configuration selection that aren't attribute-related.  We only left the not consumable failure case in `DefaultProjectDependency` different because it is due to be removed in Gradle 9.

> I feel we add more complexity to the handling of errors like this by needing to wire a resolution handler to the failure location

This type is created by a factory that already injects other resources, so it doesn't seem like a big deal to inject one more.  It's just moving some of this code to a new method there.  I think keeping a central failure location with similar call paths is much simpler than diverging how this one case is processed.

> I'm not sure I really see the value of using a failure handler here, since I can't imagine any situation where there would need to be any custom handling done here.

I disagree, I think we could easily find scenarios here.  

For instance, I've had the idea for a while that single-character transposition errors are fairly common and hard to notice by devs, but easy for Gradle.  This applies to configuration names.  Instead of just saying `'recievedMessageContainer' doesn't exist` it would be way better to say `'recievedMessageContainer' doesn't exist, did you mean 'receivedMessageContainer'?`.  Also I don't know if there's any special cases around `default` we would want to handle.

I think the point is we don't need to decide immediately, we can continue to discover new scenarios later so long as we route failures through the handler.  

I'm happy to discuss more synchronously, but I feel very strongly that unless this code is due to be removed soon, it's worth changing.

Reply:
Wired through the resolution failure handler, but did not use a failure described since this failure message is static and not dynamically configured.